zkevm-circuits/src/evm_circuit/step.rs

```diff
@@ -0,0 +1,555 @@
+                OpcodeId::PUSH12,
+                OpcodeId::PUSH13,
+                OpcodeId::PUSH14,
+                OpcodeId::PUSH15,
+                OpcodeId::PUSH16,
+                OpcodeId::PUSH17,
+                OpcodeId::PUSH18,
+                OpcodeId::PUSH19,
+                OpcodeId::PUSH20,
+                OpcodeId::PUSH21,
+                OpcodeId::PUSH22,
+                OpcodeId::PUSH23,
+                OpcodeId::PUSH24,
+                OpcodeId::PUSH25,
+                OpcodeId::PUSH26,
+                OpcodeId::PUSH27,
+                OpcodeId::PUSH28,
+                OpcodeId::PUSH29,
+                OpcodeId::PUSH30,
+                OpcodeId::PUSH31,
+                OpcodeId::PUSH32,
+            ],
+            Self::DUP => vec![
+                OpcodeId::DUP1,
+                OpcodeId::DUP2,
+                OpcodeId::DUP3,
+                OpcodeId::DUP4,
+                OpcodeId::DUP5,
+                OpcodeId::DUP6,
+                OpcodeId::DUP7,
+                OpcodeId::DUP8,
+                OpcodeId::DUP9,
+                OpcodeId::DUP10,
+                OpcodeId::DUP11,
+                OpcodeId::DUP12,
+                OpcodeId::DUP13,
+                OpcodeId::DUP14,
+                OpcodeId::DUP15,
+                OpcodeId::DUP16,
+            ],
+            Self::SWAP => vec![
+                OpcodeId::SWAP1,
+                OpcodeId::SWAP2,
+                OpcodeId::SWAP3,
+                OpcodeId::SWAP4,
+                OpcodeId::SWAP5,
+                OpcodeId::SWAP6,
+                OpcodeId::SWAP7,
+                OpcodeId::SWAP8,
+                OpcodeId::SWAP9,
+                OpcodeId::SWAP10,
+                OpcodeId::SWAP11,
+                OpcodeId::SWAP12,
+                OpcodeId::SWAP13,
+                OpcodeId::SWAP14,
+                OpcodeId::SWAP15,
+                OpcodeId::SWAP16,
+            ],
+            Self::LOG => vec![
+                OpcodeId::LOG0,
+                OpcodeId::LOG1,
+                OpcodeId::LOG2,
+                OpcodeId::LOG3,
+                OpcodeId::LOG4,
+            ],
+            Self::CREATE => vec![OpcodeId::CREATE],
+            Self::CALL => vec![OpcodeId::CALL],
+            Self::CALLCODE => vec![OpcodeId::CALLCODE],
+            Self::RETURN => vec![OpcodeId::RETURN],
+            Self::DELEGATECALL => vec![OpcodeId::DELEGATECALL],
+            Self::CREATE2 => vec![OpcodeId::CREATE2],
+            Self::STATICCALL => vec![OpcodeId::STATICCALL],
+            Self::REVERT => vec![OpcodeId::REVERT],
+            Self::SELFDESTRUCT => vec![OpcodeId::SELFDESTRUCT],
+            _ => vec![],
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct StepState<F> {
+    pub(crate) execution_state: Vec<Cell<F>>,
+    pub(crate) rw_counter: Cell<F>,
+    pub(crate) call_id: Cell<F>,
+    pub(crate) is_root: Cell<F>,
+    pub(crate) is_create: Cell<F>,
+    pub(crate) opcode_source: Cell<F>,
```

Comment:
In most situation, it would be `bytecode_hash`, but for root creation call, the `opcode_source` is actually from transaction calldata, so it would be `tx_id`. I was planning to lookup different table (`bytecode_table` and `tx_table`) respectively, but I think it deserve more discussion and then decide how to do. So I leave a todo here at line 373.

https://github.com/appliedzkp/zkevm-circuits/blob/640757ff223ad6891850ad83b6d80efe1d9112d2/zkevm-circuits/src/evm_circuit/util/constraint_builder.rs#L365-L386

One approach is to just lookup different table in different condition, but the `is_code` annotation then needs to be applied in both `bytecode_table` and `tx_table`, which I am not sure if acceptable.

Another one is to have `bytecode_table` to copy calldata from `tx_table` and create a actually unused `bytecode_hash` for this calldata as `opcode_source`, then we only need to do `opcode_lookup` to `bytecode_table` and `is_code` annotation only needs to be done in `bytecode_table`. However, I'm also afraid this would blow up `bytecode_table` because root creation call could be cheap to be huge, unless EIPs like [`EIP3680`](https://eips.ethereum.org/EIPS/eip-3860) is adopted, I think it somehow opens a DOS vector to prover.

Will keep the name `opcode_source` for now, and will file an issue to explain the consideration add some comment above line 395 to link to the issue. Do you think it's enough, or we should rename right now to avoid ambiguity it?

Reply:
Issue created appliedzkp/zkevm-specs#73 and commented added in [`4229f84`](https://github.com/appliedzkp/zkevm-circuits/pull/196/commits/4229f841f47a8795345066430cc85873bd031005)